kevm-pyk/src/kevm_pyk/utils.py

```diff
@@ -102,22 +104,57 @@ def run_prover(
     counterexample_info: bool = False,
     always_check_subsumption: bool = False,
     fast_check_subsumption: bool = False,
+    max_frontier_parallel: int = 1,
 ) -> bool:
     prover: APRProver | ImpliesProver
     try:
         if type(proof) is APRProof:
-            prover = APRProver(
-                kcfg_explore,
-                execute_depth=max_depth,
-                terminal_rules=terminal_rules,
-                cut_point_rules=cut_point_rules,
-                counterexample_info=counterexample_info,
-                always_check_subsumption=always_check_subsumption,
-                fast_check_subsumption=fast_check_subsumption,
-            )
-            prover.advance_proof(proof, max_iterations=max_iterations, fail_fast=fail_fast)
+
+            def create_prover() -> APRProver:
+                assert create_kcfg_explore is not None
+                return APRProver(
+                    create_kcfg_explore(),
+                    execute_depth=max_depth,
+                    terminal_rules=terminal_rules,
+                    cut_point_rules=cut_point_rules,
+                    counterexample_info=counterexample_info,
+                    always_check_subsumption=always_check_subsumption,
+                    fast_check_subsumption=fast_check_subsumption,
+                )
+
+            if max_frontier_parallel > 1 and create_kcfg_explore is not None:
+                parallel_advance_proof(
```

Comment:
It'd be nice to abstract these implementation details using "strategy" classes with a common interface. Then here, you could pick a strategy based on input parameters, but otherwise handle the different implementations uniformly.

Reply:
Do you mean in pyk in the same module as `Proof`, or localized here to where we know we're working with `APRProof`s? Assuming the former, I'm not sure what the common interface would look like. I'm picturing something like:
```py
class ProvingStrategy(ABC):
    @abstractmethod
    def prove(self, proof: P, max_iterations: int | None = None, fail_fast: bool = False) -> None: ...


class MultithreadedProvingStrategy(ProvingStrategy):
    _create_prover: Callable[[], Prover]
    _max_workers = 1

    def __init__(self, create_prover: Callable[[], Prover], max_workers: int = 1) -> None:
        self._create_prover = create_prover
        self._max_workers = max_workers

    def prove(self, proof: P, max_iterations: int | None = None, fail_fast: bool = False) -> None:
        parallel_advance_proof(
            proof=proof,
            create_prover=self._create_prover,
            max_iterations=max_iterations,
            fail_fast=fail_fast,
            max_workers=self._max_workers,
        )


class SequentialProvingStrategy(ProvingStrategy):
    _prover: Prover

    def __init__(self, prover: Prover) -> None:
        self._prover = prover

    def prove(self, proof: P, max_iterations: int | None = None, fail_fast: bool = False) -> None:
        self._prover.advance_proof(fail_fast=fail_fast, max_iterations=max_iterations, proof=proof)
```
in pyk, and here we'd do something like
```py
#create_prover definition
if max_frontier_parallel > 1 and create_kcfg_explore is not None:
    strategy = MultithreadedProofStrategy(create_prover, max_workers)
elif kcfg_explore is not None:
    # init APRProver
    strategy = SequentialProofStrategy(prover)
...
strategy.prove(max_iterations, fail_fast)
``` 
But all this accomplishes is condensing `advance_proof` and `parallel_advance_proof` into a single call.
I'm also hesitant to add too much that relies on the current version of `parallel_advance_proof` because per discussions with @ehildenb I'm thinking we want to expand its proving loop to deal with a collection of proofs rather than a single proof (see https://github.com/runtimeverification/k/issues/4318)

Reply:
> Do you mean in `pyk` in the same module as `Proof`, or localized here to where we know we're working with `APRProof`s?

Just here. This is not a design suggestion, just a way to make the code more readable by making it explicit that the logic is about choosing between things that are interchangeable.

Reply:
I've implemented something similar in this file. Can you see if you think this version is cleaner?

Reply:
Looks pretty good!